base/compiler/ssair/passes.jl

```diff
@@ -1255,3 +1255,185 @@ function cfg_simplify!(ir::IRCode)
+    s = stmt.args[1]
+    isa(s, QuoteNode) && (s = s.value)
+    # return any(e -> s === e, funcs)
+    return true in map(e -> s === e, funcs)
+end
+
+function is_allocation(stmt::Expr)
+    isexpr(stmt, :foreigncall) || return false
+    s = stmt.args[1]
+    isa(s, QuoteNode) && (s = s.value)
+    return (s === :jl_alloc_array_1d
+         || s === :jl_alloc_array_2d
+         || s ===  :jl_alloc_array_3d
+         || s === :jl_new_array)
+end
+
+function memory_opt!(ir::IRCode)
+    compact = IncrementalCompact(ir, false)
+    uses = IdDict{Int, Vector{Int}}()
+    relevant = IdSet{Int}() # allocations
+    revisit = Int[] # potential targets for a mutating_arrayfreeze drop-in
+    maybecopies = Int[] # calls to maybecopy
+
+    function mark_escape(@nospecialize val)
+        isa(val, SSAValue) || return
+        #println(val.id, " escaped.")
+        val.id in relevant && pop!(relevant, val.id)
+    end
+
+    function mark_use(val, idx)
+        isa(val, SSAValue) || return
+        id = val.id
+        id in relevant || return
+        (haskey(uses, id)) || (uses[id] = Int[])
+        push!(uses[id], idx)
+    end
+
+    for ((_, idx), stmt) in compact
+
+        #println("idx: ", idx, " = ", stmt)
+
+        if isa(stmt, ReturnNode)
+            isdefined(stmt, :val) || continue
+            val = stmt.val
+            mark_use(val, idx)
+            continue
+
+        # check for phinodes that are possibly allocations
+        elseif isa(stmt, PhiNode)
+
+            # ensure all of the phinode values are defined
+            defined = true
+            for i = 1:length(stmt.values)
+                if !isassigned(stmt.values, i)
+                    defined = false
+                end
+            end
+
+            defined || continue
+
+            for val in stmt.values
+                if isa(val, SSAValue) && val.id in relevant
+                    push!(relevant, idx)
+                end
+            end
+        end
+
+        (isexpr(stmt, :call) || isexpr(stmt, :foreigncall)) || continue
+
+        if is_known_call(stmt, Core.maybecopy, compact)
+            push!(maybecopies, idx)
+            continue
+        end
+
+        if is_allocation(stmt)
+            push!(relevant, idx)
+            # TODO: Mark everything else here
+            continue
+        end
+
+        if is_known_call(stmt, arrayset, compact) && length(stmt.args) >= 5
+            # The value being set escapes, everything else doesn't
+            mark_escape(stmt.args[4])
+            arr = stmt.args[3]
+            mark_use(arr, idx)
+
+        elseif is_known_call(stmt, arrayref, compact) && length(stmt.args) == 4
+            arr = stmt.args[3]
+            mark_use(arr, idx)
+
+        elseif is_known_call(stmt, setindex!, compact) && length(stmt.args) == 4
+            # handle similarly to arrayset
+            val = stmt.args[3]
+            mark_escape(val)
+
+            arr = stmt.args[2]
+            mark_use(arr, idx)
+
```

Comment:
I don't think we shouldn't assume anything on generic function like `setindex!`, since it can be overwrote/overloaded anytime.